drivers/mil_passive_sonar/scripts/multilateration

```diff
@@ -57,25 +67,82 @@ class MultilaterationNode(object):
+        # Check if two samples were taken too close to each other
+        if self.last_pos is not None and np.linalg.norm(self.last_pos - origin) < self.minimum_delta_dist:
+            rospy.loginfo("Observation too close to previous, ignoring")
+            return
+        self.last_pos = origin
+
+        # Add ping to samples list
         self.vec_samples.append((origin, origin + vec))
-        if len(self.vec_samples) == self.vec_samples.maxlen:
+
+        # If there are enough samples, calculate position
+        if len(self.vec_samples) >= self.buffer_min:
             self.publish_position()
 
     def publish_position(self):
-        line_array = np.array([(p[0][0], p[0][1], p[1][0], p[1][1]) for p in self.vec_samples])
+        # Setup lines from samples
+        line_array = np.array([(np.array(p[0][0:2]), np.array(p[1][0:2])) for p in self.vec_samples])
+
+        # Calculate least squares intersection
         where = self.ls_line_intersection2d(line_array)
+
+        # If no point could be determined, don't publish
+        if where is None:
+            return
+
+        # Publish point
         ps = PointStamped()
+        ps.header.frame_id = self.global_frame
         ps.header.stamp = rospy.Time.now()
         ps.point = numpy_to_point(where)
         self.position_pub.publish(ps)
 
-    def ls_line_intersection2d(self, line_array):
+    def ls_line_intersection2d(self, lines_array):
         """
         Find the intersection of lines in the least-squares sense.
-        start - Nx3 numpy array of start points
-        end - Nx3 numpy array of end points
-        https://en.wikipedia.org/wiki/Line-line_intersection#In_two_dimensions_2
+        Filter lobs such that no lob can have more than a number of intersections with other lines within a radius of the boat
         """
+
+        line_array = np.array([[0, 0, 0, 0]])
+        line_intersecting_array = np.array([])
+
+        for line1 in lines_array:
+            line_intersections = 0
+            for line2 in lines_array:
+                if np.array_equal(line1, line2):
+                    continue
+
+                boat_pos1 = line1[0]
+                boat_pos2 = line2[0]
+
+                # Calculate the intersection of the two lines
+                # https://stackoverflow.com/a/47823499
+                t, s = np.linalg.solve(np.array([line1[1]-line1[0], line2[0]-line2[1]]).T, line2[0]-line1[0])
+                intersection = (1-t)*line1[0] + t*line1[1]
+
+                # Calculate the distance to the boat's position from the intersection
+                distance1 = np.linalg.norm(intersection - boat_pos1)
+                distance2 = np.linalg.norm(intersection - boat_pos2)
+
+                # If the distances are within the filter radius, increment the intersections count
+                if distance1 < self.filter_dist or distance2 < self.filter_dist:
+                    line_intersections += 1
+
+            if line_intersections < self.filter_dist:
```

Comment:
Should this be `self.filter_count`?

Reply:
Good catch! I've fixed this now. I guess it passed my testing since the default values are close.